---
title: "Programming and Tidy Data Analysis in R"
output: 
    html_document:
        theme: cerulean
        css: style.css
---

<br>
These are the course notes for the [Monash Data Fluency](https://www.monash.edu/data-fluency)'s "Programming and Tidy data analysis in R" course. This is an intermediate level R workshop, for people wishing to increase their fluency in R. It covers tasks that while not specifically involved in statistical analysis are necessary when working with data: loading data and getting it into a form that is easy to work with, automating repetitive tasks, identifying problems such as the need for normalization and transformation, and properly understanding the story the data wants to tell.

## Slides

* [Introduction](slides/introduction.html) ([.Rmd](slides/introduction.Rmd))

## Topics

* [Programming](topics/programming.html) ([.Rmd](topics/programming.Rmd))
* [Communicating your results](topics/communication.html) ([.Rmd](topics/communication.Rmd))
* [The Tidyverse: dplyr, ggplot2, and friends](topics/tidyverse.html) ([.Rmd](topics/tidyverse.Rmd))

## Setup

This workshop is designed to work [Posit Cloud](https://posit.cloud/). Monash users can log in to Posit Cloud with their Monash google account. You can also use R and RStudio on your own laptop.

Create a new project, either in Posit Cloud or RStudio on your own laptop, then run the R code below. This will download files and install packages used in this workshop. If using R on your own laptop, then to use [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) you will need to have installed Java, and the FastQC steps will probably need some adjustments for Windows computers (this isn't an essential part of the workshop).

```{r eval=FALSE}
# Download workshop files
download.file(
    "https://monashdatafluency.github.io/r-progtidy/r-progtidy-files.zip",
    "r-progtidy-files.zip")
unzip("r-progtidy-files.zip")

# Download FastQC software
download.file(
    "https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.8.zip",
    "fastqc_v0.11.8.zip")
unzip("fastqc_v0.11.8.zip")
system("chmod u+x FastQC/fastqc")

# Install tidyverse packages and rmarkdown
install.packages(c("tidyverse", "rmarkdown", "quarto"))

# Optional: install some further packages 
# (these are mentioned but not essential to the workshop)
install.packages(c("devtools", "usethis", "BiocManager"))
BiocManager::install("edgeR")
```

## Files

The workshop files downloaded by the setup instructions above are also available here:

* [r-progtidy-files.zip](r-progtidy-files.zip)


## See also

<!--
* [R workshop material from Monash Data Fluency](https://www.monash.edu/data-fluency/workshops)
* [Monash Data Fluency](https://www.monash.edu/data-fluency)
* [Monash Bioinformatics Platform](https://www.monash.edu/researchinfrastructure/bioinformatics)
-->

* [Monash Genomics and Bioinformatics Platform](https://www.monash.edu/researchinfrastructure/mgbp)
* [Monash eResearch training page](https://docs.erc.monash.edu/training/)
* [Introduction-to-R workshop "next steps" page](https://monashdatafluency.github.io/r-intro-2/next_steps.html)

<br>

* ["R for Data Science" book](https://r4ds.hadley.nz/)
* [Hadley Wickham's website](http://hadley.nz/) has further useful resources.
* [The official R manuals](https://cran.r-project.org/manuals.html) give a complete description of the R language.


## Authors

This course was developed by Paul Harrison and Richard Beare for [Monash Data Fluency](https://www.monash.edu/data-fluency) and the [Monash Genomics and Bioinformatics Platform](https://www.monash.edu/researchinfrastructure/mgbp).

<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="figures/CC-BY.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.

## Source code

* [GitHub repository](https://github.com/MonashDataFluency/r-progtidy)




